File: core/tests/repos/python_core/core/scheduler.py
from .memory import MemoryPool

# vantage:
#   invariant: DeterministicOrdering
#   reason: Task execution order must be reproducible
class Scheduler:
    def __init__(self):
        self.pool = MemoryPool()
        self.tasks = []

    def spawn(self, name):
        self.pool.allocate(len(name))
        self.tasks.append(name)

    def run_all(self):
        for task in self.tasks:
            print(f"Running: {task}")

File: core/tests/repos/rust_core/src/scheduler.rs
use crate::memory::MemoryPool;

// vantage:
//   invariant: DeterministicOrdering
//   reason: Task execution order must be reproducible
pub struct Scheduler {
    pool: MemoryPool,
    tasks: Vec<String>,
}

impl Scheduler {
    pub fn new() -> Self {
        Scheduler {
            pool: MemoryPool::new(),
            tasks: Vec::new(),
        }
    }

    pub fn spawn(&mut self, name: &str) {
        self.pool.allocate(name.len());
        self.tasks.push(name.to_string());
    }

    pub fn run_all(&self) {
        for task in &self.tasks {
            println!("Running: {}", task);
        }
    }
}
